google/cloud/security/scanner/audit/iap_rules_engine.py

```diff
@@ -0,0 +1,491 @@
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (int): snapshot to load
+        """
+        super(IapRulesEngine,
+              self).__init__(
+                  rules_file_path=rules_file_path,
+                  snapshot_timestamp=snapshot_timestamp)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs):
+        """Build IapRuleBook from the rules definition file.
+
+        Args:
+            global_configs (dict): Global configurations.
+        """
+        self.rule_book = IapRuleBook(
+            global_configs,
+            self._load_rule_definitions(),
+            snapshot_timestamp=self.snapshot_timestamp)
+
+    def find_iap_violations(self, iap_resource, force_rebuild=False):
+        """Determine whether IAP-related settings violate rules.
+
+        Args:
+            iap_resource (IapResource): IAP resource data
+            force_rebuild (bool): whether to rebuild the rulebook
+
+        Returns:
+            list: RuleViolation
+        """
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+        return self.rule_book.find_violations(iap_resource)
+
+    def add_rules(self, rules):
+        """Add rules to the rule book.
+
+        Args:
+            rules (list): dicts defining the rules to add"""
+        if self.rule_book is not None:
+            self.rule_book.add_rules(rules)
+
+
+class IapRuleBook(bre.BaseRuleBook):
+    """The RuleBook for enforcing IAP policy on backend service resources."""
+
+    def __init__(self, global_configs, rule_defs=None, snapshot_timestamp=None):
+        """Initialization.
+
+        Args:
+            global_configs (dict): Global configurations.
+            rule_defs (list): IAP rule definition dicts
+            snapshot_timestamp (int): Snapshot timestamp.
+        """
+        super(IapRuleBook, self).__init__()
+        self.resource_rules_map = {}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+        if snapshot_timestamp:
+            self.snapshot_timestamp = snapshot_timestamp
+        self.org_res_rel_dao = org_resource_rel_dao.OrgResourceRelDao(
+            global_configs)
+        self.project_dao = project_dao.ProjectDao(global_configs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book.
+
+        Args:
+            rule_defs (list): rule definition property dicts"""
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): rule definition properties
+            rule_index (int): index of the rule from the rule definitions,
+                              assigned automatically when the rule book is built
+        """
+
```

Comment:
could you remove the newline between """ and the first line of code in this method?